openshift-storage-libs/openshiftstoragelibs/baseclass.py

```diff
@@ -581,6 +589,115 @@ def create_heketi_volume_with_name_and_wait(
+                - persistent_volume : (bool)
+                - persistent_volume_endpoint : (str)|None
+                - persistent_volume_file : (str)|None
+                - redundancy : (int):None
+                - replica : (int)|None
+                - size : (int):None
+                - snapshot-factor : (float)|None
+                - json : (bool)
+                - secret : (str)|None
+                - user : (str)|None
+        """
+        kwargs.pop("json", None)
+        h_vols, async_obj = [], []
+
+        # Fetch heketi volume list before operation
+        h_vols_count_initial = len(heketi_volume_list(
+            self.heketi_client_node, self.heketi_server_url,
+            json=True).get('volumes'))
+
+        # Temporary replace g.run with g.async_run in heketi_volume_create func
+        # to be able to run it in background. Also, avoid parsing the output as
+        # it won't be json at that moment. Parse it after reading the async
+        # operation results.
+        def run_async(cmd, hostname, raise_on_error=True):
+            async_op = g.run_async(host=hostname, command=cmd)
+            async_obj.append(async_op)
+            return async_op
+
+        # Waiter for overall volumes
+        _waiter = Waiter(timeout=timeout, interval=interval)
+        while vol_count > 0:
+            # Change batch_amount if pvc_count < batch_amount
+            if vol_count < batch_amount:
+                batch_amount = vol_count
+            for _ in range(batch_amount):
+                with mock.patch.object(
+                        json, 'loads', side_effect=(lambda j: j)):
+                    with mock.patch.object(
+                            command, 'cmd_run', side_effect=run_async):
+                        heketi_volume_create(
+                            self.heketi_client_node, self.heketi_server_url,
+                            size, json=True, **kwargs)
+            h_vols_count_initial += batch_amount
+
+            # Wait for all the volume to be created
+            for w in _waiter:
+                h_vols_count_final = len(heketi_volume_list(
+                    self.h_node, self.h_server, json=True).get('volumes'))
+                if (h_vols_count_final == h_vols_count_initial):
+                    break
+            if w.expired:
+                # Delete all the volumes i.e created before failure
+                h_vols = heketi_volume_list(
+                    self.h_node, self.h_server, json=True).get('volumes')
+                for vol in h_vols:
+                    self.addCleanup(
+                        heketi_volume_delete,
+                        self.heketi_client_node, self.heketi_server_url, vol)
+
+                raise AssertionError(
+                    "Failed to create volumes after waiting for {} "
+                    "secs".format(timeout))
+            vol_count -= batch_amount
+
+        # Check that all background processes got exited
+        for obj in async_obj:
```

Comment:
Add this check inside  while loop to check async objects for batch

Reply:
Done